File: Stack Problem/394. Decode String/solution.ts
function decodeString(s: string): string {
    const stack: (string | number)[] = [];
    let currentNum = 0;
    let currentString = '';

    for (const char of s) {
        if (!isNaN(parseInt(char))) {
            currentNum = currentNum * 10 + parseInt(char);
        } else if (char === '[') {
            stack.push(currentString);
            stack.push(currentNum);
            currentString = '';
            currentNum = 0;
        } else if (char === ']') {
            const num = stack.pop() as number;
            const prevString = stack.pop() as string;
            currentString = prevString + currentString.repeat(num);
        } else {
            currentString += char;
        }
    }

    return currentString;
}

File: Stack Problem/394. Decode String/solution.py
class Solution:
    def decodeString(self, s: str) -> str:
        stack = []
        current_num = 0
        current_string = ''
        
        for char in s:
            if char.isdigit():
                current_num = current_num * 10 + int(char)
            elif char == '[':
                stack.append(current_string)
                stack.append(current_num)
                current_string = ''
                current_num = 0
            elif char == ']':
                num = stack.pop()
                prev_string = stack.pop()
                current_string = prev_string + num * current_string
            else:
                current_string += char
        
        return current_string